File: evaluation/S1/java/Hard/1870-number-of-atoms.java
// Source: https://leetcode.com/problems/number-of-atoms/   |   Difficulty: Hard
//
// Problem Description:
// Given a string formula representing a chemical formula, return the count of each atom.
//
// The atomic element always starts with an uppercase character, then zero or more lowercase letters, representing the name.
//
// One or more digits representing that element's count may follow if the count is greater than 1. If the count is 1, no digits will follow.
//
//
// 	For example, "H2O" and "H2O2" are possible, but "H1O2" is impossible.
//
//
// Two formulas are concatenated together to produce another formula.
//
//
// 	For example, "H2O2He3Mg4" is also a formula.
//
//
// A formula placed in parentheses, and a count (optionally added) is also a formula.
//
//
// 	For example, "(H2O2)" and "(H2O2)3" are formulas.
//
//
// Return the count of all elements as a string in the following form: the first name (in sorted order), followed by its count (if that count is more than 1), followed by the second name (in sorted order), followed by its count (if that count is more than 1), and so on.
//
// The test cases are generated so that all the values in the output fit in a 32-bit integer.
//
// Example:
// Input: formula = "H2O"
// Output: "H2O"
// Explanation: The count of elements are {'H': 2, 'O': 1}.
//
// Constraints:
// 1 <= formula.length <= 1000
// 	formula consists of English letters, digits, '(', and ')'.
// 	formula is always valid.
//
// Helpful references (internal KB):
// - String Sorts (array, string, number, radix-sort, counting-sort)
//   • When to use: Use when sorting a large collection of fixed-length strings or integers, especially when the key range (alphabet size) is small. It is efficient for data with uniform key lengths.
//   • Idea: LSD Radix Sort sorts elements by processing digits or characters from the least significant to the most significant, using a stable counting sort for each position. It achieves a time complexity of O(W * (N + R)), where W is the key length, N is the number of items, and R is the radix.
//   • Invariants: After each pass 'd' (from 'w-1' down to '0'), the array 'a' is stably sorted based on characters from position 'd' to 'w-1'.; The 'count' array accurately reflects the frequency of each character at the current digit 'd' before cumulation.
//   • Tips: Utilize counting sort as the stable inner sorting algorithm for each digit pass.; Process characters or digits from right to left (least significant to most significant).
//   • Pitfalls: Inefficient for variable-length strings without proper padding or specialized handling.; Can be slower than comparison sorts for small input sizes or very large radix (R).
// - Balanced bracket sequences (number, counting)
//   • When to use: Use this when verifying if a sequence of delimiters is correctly nested and balanced. It is also applicable for counting or generating all possible valid sequences of a given length.
//   • Idea: This algorithm involves checking if a sequence of opening and closing delimiters is properly matched and nested, or enumerating such sequences. Validation typically uses a stack in O(N) time, while generation often involves backtracking with O(Catalan(N)) complexity.
//   • Invariants: The stack contains only opening delimiters that are yet to be matched.; The balance counter (open - close) is never negative at any point during traversal.
//   • Tips: Use a stack to track unmatched opening delimiters for validation.; Maintain a balance counter (increment for open, decrement for close) for a simpler check.
//   • Pitfalls: Forgetting to check if the stack is empty after processing the entire sequence.; Allowing a closing delimiter when no matching opening delimiter is available (empty stack or mismatch).
// - Generating all K-combinations (bitset, enumeration)
//   • When to use: When all subsets of a fixed size K need to be generated, either in lexicographical order or such that successive combinations differ by exactly one element.
//   • Idea: This algorithm enumerates all K-combinations from a set, either in lexicographical order or using a Gray code approach where adjacent combinations differ by a single element. It typically runs in O(N choose K) time, generating each combination in O(K) or O(1) amortized time.
//   • Invariants: Each generated combination contains exactly K distinct elements.; All elements within a combination are unique and from the original set.
//   • Tips: For lexicographical generation, find the rightmost element that can be incremented without exceeding its maximum possible value.; After incrementing an element, set all subsequent elements to their minimum valid values.
//   • Pitfalls: Incorrectly handling boundary conditions for the rightmost element or its maximum value.; Off-by-one errors when assigning subsequent elements after an increment.
//
// Notes:
// - Keep the given method signatures unchanged.
// - Use only Java standard library (`java.*`); no external dependencies.
// - Do not change input/output formats or return value semantics.
// - Conformance: single public class 'Solution'; no 'package' statement; place any extra types as static inner classes.
// - If you introduce any helper functions, include their full implementations in this file; no hidden helpers.

class Solution {
    public String countOfAtoms(String formula) {
        int i = 0;
        int n = formula.length();
        Deque<Map<String, Integer>> stack = new ArrayDeque<>();
        stack.push(new TreeMap<>());

        while (i < n) {
            char ch = formula.charAt(i);
            if (ch == '(') {
                stack.push(new TreeMap<>());
                i++;
            } else if (ch == ')') {
                i++;
                int start = i;
                while (i < n && Character.isDigit(formula.charAt(i))) {
                    i++;
                }
                int multiplier = start < i ? Integer.parseInt(formula.substring(start, i)) : 1;
                Map<String, Integer> top = stack.pop();
                Map<String, Integer> current = stack.peek();
                for (String atom : top.keySet()) {
                    current.put(atom, current.getOrDefault(atom, 0) + top.get(atom) * multiplier);
                }
            } else {
                int start = i;
                i++;
                while (i < n && Character.isLowerCase(formula.charAt(i))) {
                    i++;
                }
                String atom = formula.substring(start, i);
                start = i;
                while (i < n && Character.isDigit(formula.charAt(i))) {
                    i++;
                }
                int count = start < i ? Integer.parseInt(formula.substring(start, i)) : 1;
                Map<String, Integer> current = stack.peek();
                current.put(atom, current.getOrDefault(atom, 0) + count);
            }
        }

        StringBuilder result = new StringBuilder();
        Map<String, Integer> finalCount = stack.pop();
        for (String atom : finalCount.keySet()) {
            result.append(atom);
            int count = finalCount.get(atom);
            if (count > 1) {
                result.append(count);
            }
        }

        return result.toString();           
    }
}

File: evaluation/S0/c/Hard/1870-number-of-atoms.c
// Source: https://leetcode.com/problems/number-of-atoms/   |   Difficulty: Hard
//
// Problem Description:
// Given a string formula representing a chemical formula, return the count of each atom.
//
// The atomic element always starts with an uppercase character, then zero or more lowercase letters, representing the name.
//
// One or more digits representing that element's count may follow if the count is greater than 1. If the count is 1, no digits will follow.
//
//
// 	For example, "H2O" and "H2O2" are possible, but "H1O2" is impossible.
//
//
// Two formulas are concatenated together to produce another formula.
//
//
// 	For example, "H2O2He3Mg4" is also a formula.
//
//
// A formula placed in parentheses, and a count (optionally added) is also a formula.
//
//
// 	For example, "(H2O2)" and "(H2O2)3" are formulas.
//
//
// Return the count of all elements as a string in the following form: the first name (in sorted order), followed by its count (if that count is more than 1), followed by the second name (in sorted order), followed by its count (if that count is more than 1), and so on.
//
// The test cases are generated so that all the values in the output fit in a 32-bit integer.
//
// Example:
// Input: formula = "H2O"
// Output: "H2O"
// Explanation: The count of elements are {'H': 2, 'O': 1}.
//
// Constraints:
// 1 <= formula.length <= 1000
// 	formula consists of English letters, digits, '(', and ')'.
// 	formula is always valid.
//

char* countOfAtoms(char* formula) {
    int len = strlen(formula);
    int *stack = (int*)malloc((len + 1) * sizeof(int));
    int top = -1;
    int *counts = (int*)calloc(52, sizeof(int)); // 26 lowercase + 26 uppercase
    int i = 0;

    while (i < len) {
        if (formula[i] == '(') {
            stack[++top] = -1; // Marker for '('
            i++;
        } else if (formula[i] == ')') {
            i++;
            int mult = 0;
            while (i < len && formula[i] >= '0' && formula[i] <= '9') {
                mult = mult * 10 + (formula[i] - '0');
                i++;
            }
            if (mult == 0) mult = 1;

            // Pop until '(' marker
            int tempCounts[52] = {0};
            while (top >= 0 && stack[top] != -1) {
                int idx = stack[top--];
                tempCounts[idx] += 1;
            }
            top--; // Pop the '(' marker

            // Multiply and push back
            for (int j = 0; j < 52; j++) {
                if (tempCounts[j] > 0) {
                    for (int k = 0; k < mult * tempCounts[j]; k++) {
                        stack[++top] = j;
                    }
                }
            }
        } else {
            // Parse element name
            int start = i;
            i++; // Move past uppercase letter
            while (i < len && formula[i] >= 'a' && formula[i] <= 'z') {
                i++;
            }
            char elem[3] = {0};
            strncpy(elem, &formula[start], i - start);

            // Parse count
            int count = 0;
            while (i < len && formula[i] >= '0' && formula[i] <= '9') {
                count = count * 10 + (formula[i] - '0');
                i++;
            }
            if (count == 0) count = 1;

            // Map element to index
            int idx;
            if (elem[1] == '\0') { // Single letter element
                idx = elem[0] - 'A';
            } else { // Two  
                idx = 26 + (elem[0] - 'A') * 26 + (elem[1] - 'a');
            }
            for (int j = 0; j < count; j++) {
                stack[++top] = idx;
            }
        }
    }
    // Count occurrences
    for (int i = 0; i <= top; i++) {
        counts[stack[i]] += 1;
    }
    free(stack);
    // Build result string
    char *result = (char*)malloc(3000 * sizeof(char));
    result[0] = '\0';
    char buffer[20];
    for (int i = 0; i < 52; i++) {
        if (counts[i] > 0) {
            char elem[3] = {0};
            if (i < 26) {
                elem[0] = 'A' + i;
            } else {
                elem[0] = 'A' + (i - 26) / 26;
                elem[1] = 'a' + (i - 26) % 26;
            }
            strcat(result, elem);
            if (counts[i] > 1) {
                sprintf(buffer, "%d", counts[i]);
                strcat(result, buffer);
            }
        }
    }
    free(counts);
    return result;   
}
